eventmesh/LockManager.js

```diff
@@ -0,0 +1,94 @@
+'use strict';
+
+const _ = require('lodash');
+const config = require('../common/config');
+const CONST = require('./constants');
+
+const {
+  Etcd3
+} = require('etcd3');
+const client = new Etcd3({
+  hosts: config.etcd.url,
+  credentials: {
+    rootCertificate: config.etcd.ssl.ca,
+    privateKey: config.etcd.ssl.key,
+    certChain: config.etcd.ssl.crt
+  }
+});
+
+class LockManager {
+  /*
+  Lock is tracked via the key value "resource/lock/details". 
+  The value is a JSON type and the structure looks like the following.
+  {count: INT, operaitonType:STRING}
+  when a lock is taken, count value is set to 1.
+  operationType value is either "READ or "WRITE", depending on if the ongoing operation will make sync operations like bind/unbind wait for it.
+
+  This value is read and updated to acquire lock. In order to make this two different operations synchronous, we do it taking a lock on the "resource/lock" key, which would ensure synchrounous execution of the read and update operation.
+
+  Lock Algorithm looks like the following
+
+      1. Lock "resource/lock"
+      2. Check "resource/lock/details" value
+      3. if lock value > 1 then lock is already acquired by someone, 
+         unlock "resource/lock" and throw exception.
+      4. else, set the lock details as 
+         {count: 1, operationType: "READ"} 
+      5. release the lock on "resource/lock"
+  */
+  lock(resource, operationType) {
+    const lock = client.lock(resource + '/lock');
+    return lock.ttl(5).acquire()
+      .then(() => {
+        return client.get(resource + '/lock/details').json();
```

Comment:
Will json() throw error if key is not there?

Reply:
No it won't. It returns null in that case.